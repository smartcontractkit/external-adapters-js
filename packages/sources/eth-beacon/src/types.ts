export type ValidatorStatus =
  | 'pending_initialized' // When the first deposit is processed, but not enough funds are available (or not yet the end of the first epoch) to get validator into the activation queue.
  | 'pending_queued' // When validator is waiting to get activated, and have enough funds etc. while in the queue, validator activation epoch keeps changing until it gets to the front and make it through (finalization is a requirement here too).
  | 'active_ongoing' // When validator must be attesting, and have not initiated any exit.
  | 'active_exiting' // When validator is still active, but filed a voluntary request to exit.
  | 'active_slashed' // When validator is still active, but have a slashed status and is scheduled to exit.
  | 'exited_unslashed' // When validator has reached reguler exit epoch, not being slashed, and doesn't have to attest any more, but cannot withdraw yet.
  | 'exited_slashed' // When validator has reached reguler exit epoch, but was slashed, have to wait for a longer withdrawal period.
  | 'withdrawal_possible' // After validator has exited, a while later is permitted to move funds, and is truly out of the system.
  | 'withdrawal_done' // (not possible in phase0, except slashing full balance) - actually having moved funds away

export type StateId =
  | 'head' // canonical head in node's view
  | 'genesis'
  | 'finalized'
  | 'justified'
  | number // slot
  | string // hex encoded stateRoot with 0x prefix
