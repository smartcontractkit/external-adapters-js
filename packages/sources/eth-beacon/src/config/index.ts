import { Requester, util } from '@chainlink/ea-bootstrap'
import { Config } from '@chainlink/ea-bootstrap'

export const NAME = 'ETH_BEACON'

export const DEFAULT_ENDPOINT = 'balance'
export const ENV_ETHEREUM_RPC_URL = 'ETHEREUM_RPC_URL'
export const ENV_FALLBACK_RPC_URL = 'RPC_URL'
export const ENV_BATCH_SIZE = 'BATCH_SIZE'

export const DEFAULT_BATCH_SIZE = 15

export const makeConfig = (prefix?: string): Config => {
  const rpcURL = util.getRequiredEnvWithFallback(
    ENV_ETHEREUM_RPC_URL,
    [ENV_FALLBACK_RPC_URL],
    prefix,
  )
  const batchSize = parseInt(util.getEnv(ENV_BATCH_SIZE, prefix) || '') || DEFAULT_BATCH_SIZE
  const defaultConfig = Requester.getDefaultConfig(prefix)
  return {
    ...defaultConfig,
    api: {
      ...defaultConfig.api,
      baseURL: rpcURL,
    },
    defaultEndpoint: DEFAULT_ENDPOINT,
    adapterSpecificParams: {
      batchSize: batchSize,
    },
  }
}
