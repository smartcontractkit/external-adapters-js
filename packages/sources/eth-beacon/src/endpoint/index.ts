import type { TInputParameters as ValidatorInputParameters } from './validator'

export type TInputParameters = ValidatorInputParameters

export * as validator from './validator'
