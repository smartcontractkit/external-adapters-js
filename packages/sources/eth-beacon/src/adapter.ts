import { Builder } from '@chainlink/ea-bootstrap'
import {
  ExecuteWithConfig,
  ExecuteFactory,
  AdapterRequest,
  APIEndpoint,
} from '@chainlink/ea-bootstrap'
import { EthBeaconConfig, makeConfig } from './config'
import * as endpoints from './endpoint'

export const execute: ExecuteWithConfig<EthBeaconConfig, endpoints.TInputParameters> = async (
  request,
  context,
  config,
) => {
  return Builder.buildSelector<EthBeaconConfig, endpoints.TInputParameters>(
    request,
    context,
    config,
    endpoints,
  )
}

export const endpointSelector = (
  request: AdapterRequest,
): APIEndpoint<EthBeaconConfig, endpoints.TInputParameters> =>
  Builder.selectEndpoint<EthBeaconConfig, endpoints.TInputParameters>(
    request,
    makeConfig(),
    endpoints,
  )

export const makeExecute: ExecuteFactory<EthBeaconConfig, endpoints.TInputParameters> = (
  config,
) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}
