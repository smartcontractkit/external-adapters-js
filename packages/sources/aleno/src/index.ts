import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { price } from './endpoint'

export const adapter = new Adapter({
  defaultEndpoint: price.name,
  name: 'ALENO',
  config,
  endpoints: [price],
  rateLimiting: {
    tiers: {
      default: {
        rateLimit1m: 30,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
