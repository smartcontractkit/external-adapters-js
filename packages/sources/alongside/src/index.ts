import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { collateral } from './endpoint'
import { config } from './config'

export const adapter = new Adapter({
  defaultEndpoint: collateral.name,
  name: 'ALONGSIDE',
  config,
  endpoints: [collateral],
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
