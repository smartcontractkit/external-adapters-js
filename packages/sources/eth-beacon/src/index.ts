import { expose } from '@chainlink/ea-bootstrap'
import { makeExecute, endpointSelector } from './adapter'
import * as endpoints from './endpoint'
import { makeConfig, NAME } from './config'
import { envDefaultOverrides } from './config/envDefaultOverrides'

const adapterContext = { name: NAME, envDefaultOverrides }

const { server } = expose(adapterContext, makeExecute(), undefined, endpointSelector)
export { NAME, makeExecute, makeConfig, server, endpoints }
