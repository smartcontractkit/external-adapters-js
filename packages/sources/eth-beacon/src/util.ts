import { endpointSelector, makeExecute } from './adapter'
import {
  AdapterData,
  AdapterResponseData,
  makeMiddleware,
  withMiddleware,
} from '@chainlink/ea-bootstrap'
import { AdapterContext, Config } from '@chainlink/ea-bootstrap'
import * as endpoints from './endpoint'

export function requestSelf(
  context: AdapterContext,
  id: string,
  input: endpoints.TInputParameters & { endpoint: string },
): Promise<AdapterResponseData<AdapterData>> {
  const execute = makeExecute()
  const options = {
    data: input,
    method: 'post',
    id,
  }
  return new Promise((resolve, reject) => {
    const middleware = makeMiddleware<Config, endpoints.TInputParameters>(
      execute,
      undefined,
      endpointSelector,
    )
    withMiddleware<endpoints.TInputParameters>(execute, context, middleware)
      .then((executeWithMiddleware) => {
        executeWithMiddleware(options, context)
          .then((value) => resolve(value.data))
          .catch(reject)
      })
      .catch((error) => reject(error))
  })
}
