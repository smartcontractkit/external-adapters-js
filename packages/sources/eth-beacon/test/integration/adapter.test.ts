import { AdapterRequest } from '@chainlink/ea-bootstrap'
import { SuperTest, Test } from 'supertest'
import { server as startServer } from '../../src'
import { mockBalanceSuccess, mockBalanceWithStatusSuccess } from './fixtures'
import { setupExternalAdapterTest, SuiteContext } from '@chainlink/ea-test-helpers'
import process from 'process'

describe('execute', () => {
  const id = '1'
  const context: SuiteContext = {
    req: null,
    server: startServer,
  }

  const envVariables = {
    CACHE_ENABLED: 'false',
    ETHEREUM_RPC_URL: process.env.ETHEREUM_RPC_URL || 'http://localhost:3500',
  }

  setupExternalAdapterTest(envVariables, context)

  describe('balance api', () => {
    it('should return success', async () => {
      const data: AdapterRequest = {
        id,
        data: {
          result: [
            {
              address:
                '0x8bdb63ea991f42129d6defa8d3cc5926108232c89824ad50d57f49a0310de73e81e491eae6587bd1465fa5fd8e4dee21',
            },
            {
              address:
                '0xb672b5976879c6423ad484ba4fa0e76069684eed8e2a8081f6730907f3618d43828d1b399d2fd22d7961824594f73462',
            },
          ],
        },
      }

      mockBalanceSuccess()

      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })

    it('should return success with validator that is not on the beacon chain', async () => {
      const data: AdapterRequest = {
        id,
        data: {
          result: [
            {
              address:
                '0x8bdb63ea991f42129d6defa8d3cc5926108232c89824ad50d57f49a0310de73e81e491eae6587bd1465fa5fd8e4dee21',
            },
            {
              address:
                '0xb672b5976879c6423ad484ba4fa0e76069684eed8e2a8081f6730907f3618d43828d1b399d2fd22d7961824594f73462',
            },
            {
              address:
                '0x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
            },
          ],
          validatorStatus: ['active'],
        },
      }

      mockBalanceWithStatusSuccess()

      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })
})
