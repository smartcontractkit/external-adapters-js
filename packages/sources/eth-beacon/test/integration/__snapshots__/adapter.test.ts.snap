// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute balance api should return success 1`] = `
{
  "data": {
    "result": [
      {
        "address": "0x8bdb63ea991f42129d6defa8d3cc5926108232c89824ad50d57f49a0310de73e81e491eae6587bd1465fa5fd8e4dee21",
        "balance": "32081209325",
      },
      {
        "address": "0xb672b5976879c6423ad484ba4fa0e76069684eed8e2a8081f6730907f3618d43828d1b399d2fd22d7961824594f73462",
        "balance": "32067790944",
      },
    ],
  },
  "jobRunID": "1",
  "result": [
    {
      "address": "0x8bdb63ea991f42129d6defa8d3cc5926108232c89824ad50d57f49a0310de73e81e491eae6587bd1465fa5fd8e4dee21",
      "balance": "32081209325",
    },
    {
      "address": "0xb672b5976879c6423ad484ba4fa0e76069684eed8e2a8081f6730907f3618d43828d1b399d2fd22d7961824594f73462",
      "balance": "32067790944",
    },
  ],
  "statusCode": 200,
}
`;

exports[`execute balance api should return success with validator that is not on the beacon chain 1`] = `
{
  "data": {
    "result": [
      {
        "address": "0x8bdb63ea991f42129d6defa8d3cc5926108232c89824ad50d57f49a0310de73e81e491eae6587bd1465fa5fd8e4dee21",
        "balance": "32081209325",
      },
      {
        "address": "0xb672b5976879c6423ad484ba4fa0e76069684eed8e2a8081f6730907f3618d43828d1b399d2fd22d7961824594f73462",
        "balance": "32067790944",
      },
      {
        "address": "0x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
        "balance": "0",
      },
    ],
  },
  "jobRunID": "1",
  "result": [
    {
      "address": "0x8bdb63ea991f42129d6defa8d3cc5926108232c89824ad50d57f49a0310de73e81e491eae6587bd1465fa5fd8e4dee21",
      "balance": "32081209325",
    },
    {
      "address": "0xb672b5976879c6423ad484ba4fa0e76069684eed8e2a8081f6730907f3618d43828d1b399d2fd22d7961824594f73462",
      "balance": "32067790944",
    },
    {
      "address": "0x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "balance": "0",
    },
  ],
  "statusCode": 200,
}
`;
